fem/multigrid.cpp

```diff
@@ -60,187 +45,232 @@ Multigrid::~Multigrid()
    {
-      if (ownedProlongations[i])
+      const int n = operators[i]->Height();
+      for (int j = 0; j < X.NumCols(); ++j)
       {
-         delete prolongations[i];
+         X(i, j) = new Vector(n);
+         Y(i, j) = new Vector(n);
+         R(i, j) = new Vector(n);
+         Z(i, j) = new Vector(n);
       }
    }
+}
 
-   operators.DeleteAll();
-   smoothers.DeleteAll();
-   prolongations.DeleteAll();
-   X.DeleteAll();
-   Y.DeleteAll();
-   R.DeleteAll();
-   Z.DeleteAll();
+void MultigridBase::EraseVectors() const
+{
+   for (int i = 0; i < X.NumRows(); ++i)
+   {
+      for (int j = 0; j < X.NumCols(); ++j)
+      {
+         delete X(i, j);
+         delete Y(i, j);
+         delete R(i, j);
+         delete Z(i, j);
+      }
+   }
 }
 
-void Multigrid::AddLevel(Operator* opr, Solver* smoother, bool ownOperator,
-                         bool ownSmoother)
+void MultigridBase::AddLevel(Operator* op, Solver* smoother,
+                             bool ownOperator, bool ownSmoother)
 {
-   operators.Append(opr);
+   height = op->Height();
+   width = op->Width();
+   operators.Append(op);
    smoothers.Append(smoother);
    ownedOperators.Append(ownOperator);
    ownedSmoothers.Append(ownSmoother);
-   width = opr->Width();
-   height = opr->Height();
-
-   X.Append(new Vector(height));
-   *X.Last() = 0.0;
-   Y.Append(new Vector(height));
-   *Y.Last() = 0.0;
-   R.Append(new Vector(height));
-   *R.Last() = 0.0;
-   Z.Append(new Vector(height));
-   *Z.Last() = 0.0;
 }
 
-int Multigrid::NumLevels() const { return operators.Size(); }
-
-int Multigrid::GetFinestLevelIndex() const { return NumLevels() - 1; }
-
-const Operator* Multigrid::GetOperatorAtLevel(int level) const
-{
-   return operators[level];
-}
-
-Operator* Multigrid::GetOperatorAtLevel(int level)
-{
-   return operators[level];
-}
-
-const Operator* Multigrid::GetOperatorAtFinestLevel() const
-{
-   return GetOperatorAtLevel(operators.Size() - 1);
-}
-
-Operator* Multigrid::GetOperatorAtFinestLevel()
-{
-   return GetOperatorAtLevel(operators.Size() - 1);
-}
-
-Solver* Multigrid::GetSmootherAtLevel(int level) const
-{
-   return smoothers[level];
-}
-
-Solver* Multigrid::GetSmootherAtLevel(int level)
-{
-   return smoothers[level];
-}
-
-void Multigrid::SetCycleType(CycleType cycleType_, int preSmoothingSteps_,
-                             int postSmoothingSteps_)
+void MultigridBase::SetCycleType(CycleType cycleType_, int preSmoothingSteps_,
+                                 int postSmoothingSteps_)
 {
    cycleType = cycleType_;
    preSmoothingSteps = preSmoothingSteps_;
    postSmoothingSteps = postSmoothingSteps_;
 }
 
-void Multigrid::Mult(const Vector& x, Vector& y) const
+void MultigridBase::Mult(const Vector& x, Vector& y) const
 {
-   MFEM_ASSERT(NumLevels() > 0, "");
-   *X.Last() = x;
-   *Y.Last() = 0.0;
-   Cycle(GetFinestLevelIndex());
-   y = *Y.Last();
+   Array<Vector*> X_(1), Y_(1);
+   X_[0] = const_cast<Vector*>(&x);
```

Comment:
I don't like the const_cast but there is probably no other way to do it. Or is it maybe possible to use an Array of const Vector* for X_ in the method for multiple rhs?

Reply:
Fixed in [b15b188](https://github.com/mfem/mfem/pull/3075/commits/b15b188c448d4a0086d5357fa1c376d2681dff05). I wasn't sure if this would work since I know compilers can sometimes not like `const` objects in STL vectors, but it seems to work fine here. Thanks!